core/amber/src/main/scala/edu/uci/ics/texera/web/SyncableState.scala

```diff
@@ -0,0 +1,61 @@
+package edu.uci.ics.texera.web
+
+import rx.lang.scala.Subscription
+
+import java.util.concurrent.locks.ReentrantLock
+import scala.collection.mutable
+
+class SyncableState[T](defaultStateGen: => T) {
+
+  private var state: T = defaultStateGen
+  private var isModifying = false
+  private var callbackId = 0
+  private var onChangedCallbacks = mutable.HashMap[Int, (T, T) => Unit]()
+  private val lock = new ReentrantLock()
+
+  def getStateThenConsume[X](next: T => X): X = {
+    withLock {
+      next(state)
+    }
+  }
+
+  def acquireLock(): Unit = lock.lock()
+  def releaseLock(): Unit = lock.unlock()
+
+  private def withLock[X](code: => X): X = {
+    acquireLock()
+    val result = code
```

Comment:
what happens if `code()` breaks?

Reply:
it throws an exception and the caller is responsible to handle it.

Reply:
my point is if `code()` breaks, this will result in a deadlock because the lock is never released.

Reply:
added try-catch-finally to withLock